### PR TITLE
upgrade to sbt 0.13.13-RC1

### PR DESCRIPTION
**Pull request by SethTisue**
not to be merged until 0.13.13 final is out (and we upgrade
to it instead) -- putting this out there to help test sbt

### CODE REVIEW

**Comment by SethTisue** [comment]
test run: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/311/consoleFull
fyi @eed3si9n @dwijnand

**Comment by dwijnand** [comment]
Thank you.

**Comment by SethTisue** [comment]
extraction failed, several instances of

```
[specs2] java.lang.NoSuchMethodError: org.eclipse.jgit.storage.file.FileRepositoryBuilder.build()Lorg/eclipse/jgit/storage/file/FileRepository;
[specs2]    at com.typesafe.sbt.git.JGit$.apply(JGit.scala:85)
```

**Comment by eed3si9n** [comment]
https://github.com/SethTisue/community-builds/blob/master/common-2.11.x.conf#L120 says:

``` yaml
  // TODO move back to master after adding scalaz-stream to the community build.
  specs2-ref                   : "etorreborre/specs2.git#8fa92f3f9dac61ba39e169b83c091c4d5b15cb8c"
```

That's from 2014 https://github.com/etorreborre/specs2/blob/8fa92f3f9dac61ba39e169b83c091c4d5b15cb8c/project/plugins.sbt#L7

``` scala
addSbtPlugin("com.typesafe.sbt" % "sbt-git" % "0.6.1")
```

This uses JGit 2.2 (https://github.com/sbt/sbt-git/blob/v0.6.1/build.sbt#L9):

```
libraryDependencies += "org.eclipse.jgit" % "org.eclipse.jgit.pgm" % "2.2.0.201212191850-r"
```

I've matched Giter8's JGit (https://github.com/foundweekends/giter8/blob/0.7.x/project/Dependencies.scala#L17) with the latest sbt-git

``` scala
  val jgit = "org.eclipse.jgit" % "org.eclipse.jgit.pgm" % "3.7.0.201502260915-r"
```

**Comment by SethTisue** [comment]
> https://github.com/SethTisue/community-builds/blob/master/common-2.11.x.conf#L120

(note that's not the relevant branch/file, what's relevant is common.conf on the 2.11.x branch)

so, would using a newer sbt-git help, or what? let's try again with a newer version of sbt-git in the specs2 build: https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/312/consoleFull (404 til Jenkins catches up)

**Comment by eed3si9n** [comment]
Using the latest sbt-git should help.

**Comment by dwijnand** [comment]
@SethTisue I think you can unpin the sbt ref too, this should've fixed it https://github.com/sbt/sbt/pull/1531

```
  // change from https://github.com/sbt/sbt/pull/1509 broke building sbt with Scala 2.11
  // this problem is tracked in https://github.com/sbt/sbt/issues/1523
  // for now, we stick to fixed sha1 of sbt right before the merge of #1509
  sbt-ref                      : "sbt/sbt.git#0b2f2ae5a8ab2b082b6e15b196d671a4c18cc9f2"
```
- from https://github.com/SethTisue/community-builds/blob/sbt-0.13.13/common.conf#L79-L82

**Comment by dwijnand** [comment]
That'll fix the sbt-git version in sbt/sbt

**Comment by SethTisue** [comment]
> I think you can unpin the sbt ref too

I'll try it, but note that the sbt we build in the community build doesn't actually get used for anything. we're just checking to see if it builds. (& at the moment anyway, `run-tests: false`)

**Comment by SethTisue** [comment]
```
[info] Project specs2------------------: EXTRACTION OK, but could not proceed
```

ok, so I just need to deal with the other projects then

**Comment by SethTisue** [comment]
```
[lift-json:error] /home/jenkins/workspace/scala-2.11.x-integrate-community-build/target-0.9.5/extraction/7c65e473b21706b4d5696d24bdc280d6621e0ba8/projects/52b88d657a64282461d253213501af792da9ec97/project/plugins.sbt:1: error: value addPluginResolvers is not a member of object sbt.DefaultOptions
[lift-json:error] DefaultOptions.addPluginResolvers
[lift-json:error]                ^
[lift-json] [error] Type error in expression
```

we have lift-json frozen at `3.0-M2-release`, perhaps we just need to move to something newer

**Comment by dwijnand** [comment]
> I'll try it, but note that the sbt we build in the community build doesn't actually get used for anything. we're just checking to see if it builds. (& at the moment anyway, run-tests: false)

I realise.

It's actually quite a bizarre situation, no-one uses a Scala 2.11 build of sbt, there is no community build for 2.10, and Scala 2.11 isn't being developed...

**Comment by SethTisue** [comment]
> I realise.

(sorry, I actually deleted my comment after realizing it wasn't apropos)

**Comment by dwijnand** [comment]
> [lift-json:error] /home/jenkins/workspace/scala-2.11.x-integrate-community-build/target-0.9.5/extraction/7c65e473b21706b4d5696d24bdc280d6621e0ba8/projects/52b88d657a64282461d253213501af792da9ec97/project/plugins.sbt:1: error: value addPluginResolvers is not a member of object sbt.DefaultOptions
> [lift-json:error] DefaultOptions.addPluginResolvers
> [lift-json:error]                ^
> [lift-json] [error] Type error in expression

https://github.com/sbt/sbt/pull/2749

File: common.conf [hunk 1]
```diff
@@ -225,6 +221,8 @@ build += {
     //[info] [error]  ClassNotFoundException: : scala.util.Left  (Formats.scala:223)
     // !!!: probably we should investigate
     extra.run-tests: false
+    // with 0.13.13-RC1 I hit https://github.com/sbt/sbt/pull/2749
+    extra.sbt-version: "0.13.12"
```
Comment (dwijnand) [inline common.conf:229]: oh wow, didn't think/know you could do that

Comment (SethTisue) [reply common.conf:229]: ha, well, it didn't actually work, it still extracted with 0.13.13-RC1, I guess `sbt-version` only controls the version used during building

Comment (SethTisue) [reply common.conf:229]: er no wait it did work. or did it? I'm trying to do too many things at once...

Comment (SethTisue) [reply common.conf:229]: yeah, never mind, we're good

**Comment by SethTisue** [comment]
yay extraction succeeded at https://scala-ci.typesafe.com/job/scala-2.11.x-integrate-community-build/316/consoleFull, now we'll see how the builds go

**Comment by SethTisue** [comment]
the scala-arm failure is presumably unrelated to the sbt version bump.

as for the failure of sbt to build:

```
[sbt] java.util.NoSuchElementException: None.get
[sbt]   at scala.None$.get(Option.scala:313)
[sbt]   at scala.None$.get(Option.scala:311)
[sbt]   at bintray.Prompt$.apply(Prompt.scala:12)
[sbt]   at bintray.Bintray$.bintray$Bintray$$requestCredentials(Bintray.scala:102)
[sbt]   at bintray.Bintray$$anonfun$ensuredCredentials$2.apply(Bintray.scala:64)
[sbt]   at bintray.Bintray$$anonfun$ensuredCredentials$2.apply(Bintray.scala:60)
[sbt]   at scala.util.Either.fold(Either.scala:98)
[sbt]   at bintray.Bintray$.ensuredCredentials(Bintray.scala:60)
```

this looks like the sort of thing that would probably be a simple fix to the sbt build?

**Comment by dwijnand** [comment]
Is dbuild trying to publish sbt?

Is this the first time that that issue has come up? I feel like it would've come up before in other builds, it's a bit of a known issue with bintray (that and the "Missing bintray credentials" spam)

I'm not sure what @eed3si9n thinks, but I'm almost inclined to say you can drop sbt 0.13 from the community build for 2.11 - as I said earlier I'm not sure how much sense it makes to test.

**Comment by eed3si9n** [comment]
> I'm not sure what @eed3si9n thinks, but I'm almost inclined to say you can drop sbt 0.13 from the community build for 2.11 - as I said earlier I'm not sure how much sense it makes to test.

It depends on what the community build is testing for whom. Note that sbt contains source-distributed compiler bridge because we are using Scala compiler as a library, calling unpromised methods. It would make sense to test at least that part of sbt on any actively maintained branches so we can catch breakage early.

**Comment by dwijnand** [comment]
If this were sbt 1.0 on Scala 2.12, both in active development, I would say absolutely keep it.

But it's sbt 0.13 on Scala 2.11, which are both in maintenance mode (+/-), and also a combination that isn't used anyways.

And it's been pinned for 2 years, and now, unpinned, it's broken, because of bintray-sbt reasons.

**Comment by SethTisue** [comment]
Well, what about adding sbt 1.0 instead? Would it be added to the 2.11 community build, or the 2.12 community build, or both?

**Comment by dwijnand** [comment]
Whichever version of Scala it ends up being built with (currently 2.11, but I want my trailing commas, so hopefully 2.12 xD).

But I think we should wait for either 1.0.0, or some release candidate version.

**Comment by SethTisue** [comment]
> But I think we should wait for either 1.0.0, or some release candidate version

why? we often track development branches of included projects.

**Comment by dwijnand** [comment]
We expect to make breaking changes across the sbt modules barrier before 1.0.

Wouldn't that be a problem for the community build?

**Comment by SethTisue** [comment]
> We expect to make breaking changes across the sbt modules barrier before 1.0.
> 
> Wouldn't that be a problem for the community build?

why? I'm talking about the version of sbt that we build — merely to make sure it builds (and perhaps run your tests see if they pass) — not the version of sbt we use to build anything else

**Comment by dwijnand** [comment]
sbt 1.0 is broken down into modules: http://www.scala-sbt.org/0.13/docs/Modularization.html

Namely:
- https://github.com/sbt/sbt
- https://github.com/sbt/launcher
- https://github.com/sbt/zinc
- https://github.com/sbt/librarymanagement
- https://github.com/sbt/util
- https://github.com/sbt/io
- https://github.com/sbt/serialization (this might go away)

.. and a bunch of other dependencies

The HEAD of the 1.0 branches for all these might not all work properly together in a dbuild, as we shuffle/remove/hide different parts of their API, on our way to preparing for 1.0.

**Comment by SethTisue** [comment]
I see. In that case, it sounds like we ought to use tags rather than HEAD. is there a working set of tags we could use? (And then move forward from time to time?) Or is everything just in too much flux right now?

**Comment by dwijnand** [comment]
It all works ATM, but at some point we're going to start the flux.

**Comment by SethTisue** [comment]
> It all works ATM

Then tag the present moment, and we're good to go

**Comment by SethTisue** [comment]
I'm pushing on this a bit because the sbt codebase is a hugely important codebase. It's one of the codebases we reallllllly want to know if Scalac changes are breaking.

**Comment by dwijnand** [comment]
Sounds good, good thinking.

We have this tag https://github.com/sbt/sbt/releases/tag/v1.0.0-M4 and that depends on tagged versions of the modules.

**Comment by SethTisue** [comment]
maybe the weird bintray error is https://github.com/typesafehub/dbuild/issues/158 (and perhaps using a newer version of the plugin would fix it)? I don't have time to look into it right now

**Comment by SethTisue** [comment]
closing for now. to be revived at some later date

**Comment by SethTisue** [comment]
dbuild 0.9.7-RC1 is supposed to fix https://github.com/typesafehub/dbuild/issues/158, and the community build is now using it, so the door might be open again to re-add some version of sbt

**Pull request closed without merging**